### ISSUE TITLE
[question] CMakeToolchain + missing FORCE on cache variables?

### ISSUE THREAD

**Issue by harsszegi** [conan-io/conan#13006]
### What is your question?

Hi,

I don't fully understand how CMakeToolchain generates the "cache" variables.
We have CMakeLists.txts which define option()s /set(CACHE) before project() call.
We also generate the CMake toolchain files. At the end of the day (CMakeToolchain/_template) the 
variables get set as set(name "value" CACHE STRING "comment").
We also add the package options to the generated CMake toolchain file.
As the variables are set without FORCE it means that if the given package option value was set by the given
option() / set(CACHE) before the project() call in the main CMakeLists.txt file, then the generated toolchain file
will not be overwriting this.

Any hint on how to change this behaviour? Could could the variables generated by the CMakeToolchain be
forced?
Thanks, 

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

**Comment by jcar87** [comment]
Hi @harsszegi, thanks for opening this issue.

May I ask which specific variables are being defined in your project's `CMakeLists.txt` _before_ the first call to `project()`?

In general, unless there is good reason for it (for example, variables that affect where to locate the compiler, and other things initialised during the first `project()` call) - projects should define these variables _after_ the first call to project. 

When you instantiate a `CMakeToolchain` there are two types of variables you can define, via the `variables` and `cache_variables` attributes:
*  `variables` are for variables that one would normally define in a CMake toolchain. These will be loaded during the first `project()` call, following the order defined by CMake. 
* `cache_variables` are for variables that one would normally provide in the command line with `-D` flags. These will not be output in the generated `conan_toolchain.cmake`, but rather, Conan keeps track of these to pass them when invoking CMake. When passed as `-D`, they will take precedence over the default value of any variables defined via `option()` or `set(xx CACHE)`, which I believe is the behaviour you are after.

In your `generate()` method, could you confirm if changing these to `cache_variables`  addresses your issue? 

Thanks!

**Comment by harsszegi** [comment]
Hi @jcar8,

these are custom package specific parameters, however they are used in our custom CMake callbacks (CMAKE_PROJECT_INCLUDE_*) which are executed before the project() call and use / check the value of the option (if being left at its default value or overridden manually).

I have tried used the cache_variables approach, but I have serious issues with that. Let me explain in details.
We use the generated toolchain file "occasionally" for users who do not want to build with Conan at all, but on their own. This is done via executing "conan install" in CMake time within a CMAKE_PROJECT_INCLUDE_BEFORE callback and via using a custom CMakePreset, which:
a.) prepares the dependencies based upon the selected profile by the given CMakePreset target
b.) sets the generated conan_toolchain.cmake as CMAKE_TOOLCHAIN_FILE before project() has been executed
With b.) we are sure that CMake will pick up the proper compiler / settings etc., as that is executed before the project()
call is executed hence, it is set before the toolchain detection is executed by CMake.

But! This use-case is only executed - as I wrote - for users not planning to use Conan to build the package. For the rest
"conan export, conan install, conan build, conan package, conan export-pkg, conan test , etc." is being executed.
And that means that the generated toolchain is not being used at all, we are simply building with the "cmake_paths"
generator.

However with the offered "cache_variables" approach CMakeToolchain.generate() becomes intrusive. It will affect the
regular Conan based workflow as well, as the "-Ds" are passed to the regular "conan build" step no matter what, so there
is absolutely no way to control "when" (whether the generated toolchain is being used at all or not) to activate the 
"-Ds". So it appears that our separated two different use cases become somewhat "mixed".

Anyhow thanks for the info, I'm closing this for now, most probably we'll 'sed' the generated toolchain file for now.
